zilliqa/src/db.rs

```diff
@@ -584,6 +726,63 @@ impl Db {
     }
 }
 
+pub fn snapshot_block_with_state<P: AsRef<Path> + Debug>(
+    block: &Block,
+    parent: &Block,
+    state_trie_storage: Arc<TrieStorage>,
+    shard_id: u64,
+    output: P,
+) -> Result<()> {
+    // quick sanity check
+    if block.parent_hash() != parent.hash() {
+        return Err(anyhow!(
+            "Parent block parameter must match the snapshot block's parent hash"
+        ));
+    }
+
+    // Consider if we need better error handling here, esp. what to do if the file already exists?
+    let outfile = File::create_new(output.as_ref().join("snapshot.txt"))?;
+    let mut writer = BufWriter::with_capacity(8192 * 1024, outfile); // 8 MiB chunks
+
+    // write the block...
+    writer.write_all(hex::encode(bincode::serialize(&block)?).as_bytes())?; // TODO: better serialization (#1007)
+    writer.write_all(b"\n")?;
+    // and its parent, to keep the qc tracked
+    writer.write_all(hex::encode(bincode::serialize(&parent)?).as_bytes())?; // TODO: better serialization (#1007)
+    writer.write_all(b"\n")?;
+    // write our shard id, for sanity checking
+    writer.write_all(shard_id.to_string().as_bytes())?;
+    writer.write_all(b"\n")?;
+
+    // then write state
+    let accounts = EthTrie::new(state_trie_storage.clone()).at_root(block.state_root_hash().into());
+    let account_storage = EthTrie::new(state_trie_storage);
+    let mut account_key_buf = [0u8; 64]; // save a few allocations, since account keys are fixed length
+
+    for (key, val) in accounts.iter() {
+        let account_storage =
+            account_storage.at_root(bincode::deserialize::<Account>(&val)?.storage_root);
+
+        // export the account itself
+        hex::encode_to_slice(key, &mut account_key_buf)?;
+        writer.write_all(&account_key_buf)?;
+        writer.write_all(b":")?;
+        writer.write_all(hex::encode(val).as_bytes())?;
+        writer.write_all(b";")?;
+
+        // now the account storage
+        for (storage_key, storage_val) in account_storage.iter() {
+            writer.write_all(hex::encode(storage_key).as_bytes())?;
+            writer.write_all(b":")?;
+            writer.write_all(hex::encode(storage_val).as_bytes())?;
+            writer.write_all(b",")?;
+        }
+        writer.write_all(b"\n")?;
+    }
+    writer.flush()?;
```

Comment:
How do I tell when you've finished writing the snapshot? (If I were you, I'd write snapshots to a temporary filename, then rename when done - this ensures that the name does not become available until the contents are valid - you might also usefully write a hash of the snapshot either to the snapshot or to a separate file to protect against errors in transit).

Reply:
Excellent point, thanks!